Local API: Add write endpoints, track local data object versions

AbeJellinek opened the pull request:
I haven't run the entire test suite locally, so we'll see what happens on CI, but this is intended to be a relatively unobtrusive set of changes to track a monotonically increasing library/data object version number.

When the local API is enabled, modifying a data object (item, collection, or search) increments the library's `clientVersion` (once per transaction), then sets the data object's `clientVersion` to the same value.

Also updated `toResponseJSON()` to return local storage properties as well as versions.

Closes #5011 (option C)

AbeJellinek committed:
Track data object versions locally

AbeJellinek committed:
Keep old default behavior (to fix other tests)

AbeJellinek committed:
modifyDataObject: Just pass tx: false manually

_saveData()'s requireTransaction() call was throwing in one test. Not
sure why!

AbeJellinek committed:
Return up-to-date storage properties, too

AbeJellinek committed:
Import XPCOMUtils

AbeJellinek committed:
Local API: Support write requests

dstillman commented on chrome/content/zotero/xpcom/data/dataObject.js:
```diff
@@ -23,6 +23,15 @@
+	lazy,
+	'shouldTrackClientVersions',
+	'extensions.zotero.httpServer.localAPI.enabled',
+);
```
Gating on the pref being enabled would mean that if you disabled the API, made a change, and re-enable it, clients would miss updates. I think we originally had some plan for dealing with a `since=` from before the earliest available version — I guess specifically for deletes, so that we could drop ancient delete-log entries — but I don't think we ever implemented that, and I don't think any clients would handle a 409 or whatever that would throw. We could do that, but it's probably fine to just drop the gate and always update clientVersion. I ran some quick benchmarks and it didn't seem like it had any noticeable effect on performance, even for 2000-item adds as separate transactions.

AbeJellinek replied:
Sure, I have no problem with that.

dstillman commented on chrome/content/zotero/xpcom/data/item.js:
```diff
@@ -6170,6 +6161,12 @@ Zotero.Item.prototype.toResponseJSONAsync = async function (options = {}) {
+		json.data.mtime = await this.attachmentModificationTime ?? null;
+		json.data.md5 = await this.attachmentHash ?? null;
+	}
+	
```
If we're adding these in the async path here, why are we changing the default toJSON() behavior to emit them? That could affect a lot of other things, and is at least untested.

AbeJellinek replied:
Oh, yeah, that was not a good change. I'll pull it out of `toJSON()`.

dstillman committed:
Reject upload filenames containing path separators

dstillman committed:
Accept API key via Authorization: Bearer header

dstillman committed:
Support multipart (params=1) file uploads

dstillman committed:
Fix binary file corruption in multipart/form-data parsing

_decodeMultipartData() trimmed each part body, mangling file content that
begins or ends with a whitespace byte (e.g., a PDF ending in a newline).

dstillman commented:
pyzotero uses `Authorization: Bearer <key>` and `params=` file uploads, so I added support for those.

dstillman commented:
(Its `attachment_simple()` function still won't work because that expects `/items/new` to work, which we don't support and probably don't want to.)

dstillman committed:
Check single-object write precondition against the object version

Not the library version

dstillman committed:
Report the local version consistently in object data

dstillman commented:
A couple more fixes. I ran a battery of tests with a ([fixed](https://github.com/urschrei/pyzotero/pull/331)) version of pyzotero, and it looks good! Just the couple issues mentioned above.

dstillman committed:
Always track client versions

Remove gating on the local API pref

AbeJellinek commented:
Great!

It occurs to me that we'll need to mention this as a breaking change for the local API - any code that caches versions returned by the local API could break, and mixing local API and web API code (as I think some people do, since the local API doesn't support writes!) becomes more complicated.

Should we add a marker indicating that versions returned in the response are local, so API consumers know whether they're talking an updated version of Zotero?

AbeJellinek committed:
Extract mtime/hash fallback from toJSON() to toResponseJSON()

To avoid affecting other toJSON() consumers.

dstillman commented:
You mean because before we were returning the server version, and now we're returning a local version? Arguably caching was just broken before, because an item could change locally without the version being bumped (if it hadn't yet synced), right?

But yes, trying to read from the API and then write to the web API with that version would not go well…

AbeJellinek commented:
> Arguably caching was just broken before, because an item could change locally without the version being bumped (if it hadn't yet synced), right?

That's true, and hopefully that discouraged local API consumers from relying on it. But syncing (when set up) is frequent enough that I don't think we can count on that being the case.

It could be a header - `Zotero-Object-Version-Context`? `Zotero-Object-Version-Partition`? Or something? - that contains a random ID, persisted per client in a pref, that consumers:

- Can read from responses, cache, and compare to know whether they're looking at versions that exist in the same world as ones they received previously
- Need to send back in every write request to the local API, or their request will immediately be rejected
- CANNOT send to the web API, or it will reject their request (requires a small change in the dataserver)

AbeJellinek commented:
Or the dataserver returns its own magic value, which could just be the server hostname or something like that, and which it doesn't require but does validate if it's provided.

dstillman commented:
Well, there's a broader idea where the API includes a random identifier that allows clients to know if they're talking to the same endpoint in general, which could also be used to better support custom servers (real or development). That could be advisory or I guess something the server validates. I was assuming it would just be the former — tell syncing clients to perform a full sync and make clear to other consumers that versions aren't equivalent.

AbeJellinek commented:
`Zotero-API-Endpoint-ID`?

I guess it could just be advisory. Even if we wanted to require it in the local API, I think it would only make sense to require it for writes, and those weren't supported before anyway.

AbeJellinek commented:
But we could validate it if provided in a request - no harm in that, and it allows clients to avoid making an initial request to check the ID if they think they already know it.

dstillman commented:
`Zotero-Server-ID`?

AbeJellinek commented:
Sure.

AbeJellinek committed:
/authorize-local -> /local/authorize

AbeJellinek committed:
Add Zotero-Server-ID, require for writes, validate when provided

AbeJellinek commented:
Alright, see if those do the trick.

dstillman commented on chrome/content/zotero/xpcom/server/server_localAPI.js:
```diff
@@ -65,6 +91,147 @@ That said, there are benefits:
+		id = Zotero.Utilities.randomString(12);
+		Zotero.Prefs.set('httpServer.localAPI.serverID', id);
+	}
+	return id;
```
This has to be in the DB, not in a pref, since it's tied to the data in the database. We can store it in the `settings` table.

One issue would be DB backup restores, which could change the data out from under the consumer. If we want to handle that, we could store a DB version in a pref that gets updated monotonically along with a version stored in the DB after any DB write. At startup, check whether pref-version > DB-version and roll a new server id at that time, to force consumers to perform a full sync. Wouldn't handle all cases — someone could do a Time Machine restore that restored profile dir and data dir together — but would handle the most common ones.

AbeJellinek replied:
OK, in the settings table now.

How concerned are we about the DB restore case? I think that check could lead to some confusion, but I see why we might want it, and it would be easy to put in.

AbeJellinek committed:
Store server ID in DB settings table

AbeJellinek committed:
Rewrite doc comment

dstillman committed:
Stage uploads in temp directory until registration succeeds

Previously the upload receiver wrote directly into the attachment's
storage directory, so an abandoned or failed registration left the new
bytes in place of the existing file.

dstillman committed:
Require a precondition for keyed multi-object writes

A keyed object posted with neither a version property nor
If-Unmodified-Since-Version could silently overwrite an existing
object. Fail such entries with 428, matching the dataserver, and
return 404 for a nonzero version on a nonexistent object.

dstillman committed:
Advance clientVersion on object deletion and bulk tag removal

Deletions and tag removal left the library clientVersion unchanged, so
incremental readers could get a 304 after data changed and stale
library preconditions remained valid.

dstillman committed:
Report local versions in format=versions

format=versions returned synced versions, so unsynced objects appeared
as version 0 while JSON output and since= used their local versions.

dstillman committed:
Report the synced group version consistently in group responses

The top-level version switched to the inherited clientVersion, which
tracks the group library's contents, while data.version kept the synced
group metadata version. Group metadata isn't locally writable, so
report the synced group version in both fields, in Last-Modified-
Version, and in format=versions (keyed by group ID), and drop the
unused groups.clientVersion column. Omit Last-Modified-Version for
group lists, which have no aggregate version.

dstillman committed:
Ignore 'since' for group lists

Group metadata versions are independent of each other and of any
library version, so they don't form a valid aggregate cursor to filter
against. The dataserver likewise ignores 'since' for group lists.

dstillman commented:
Rebased, and some fixes added

dstillman commented:
Merged in 9dd17a212ed63ef5eb72fc1c04d1d21cb228b537, 77f2432103e60e2b4545b1add6ac78d9cb8a5997, a37a9e782a515c26e04e67e8e692983d94753b41
